qiskit_machine_learning/neural_networks/sampler_qnn.py

```diff
@@ -0,0 +1,296 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A Neural Network implementation based on the Sampler primitive."""
+
+from __future__ import annotations
+import logging
+from numbers import Integral
+from typing import Optional, Union, List, Tuple, Callable, cast, Iterable
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter
+from qiskit.primitives import BaseSampler
+from qiskit.algorithms.gradients import BaseSamplerGradient
+from qiskit_machine_learning.exceptions import QiskitMachineLearningError, QiskitError
+from .neural_network import NeuralNetwork
+
+logger = logging.getLogger(__name__)
+
+
+class SamplerQNN(NeuralNetwork):
+    """A Neural Network implementation based on the Sampler primitive."""
+
+    def __init__(
+        self,
+        sampler: BaseSampler,
+        circuit: QuantumCircuit,
+        input_params: List[Parameter] | None = None,
```

Comment:
Do we need List  (list) or could we go with Sequence as bit more generic.

Reply:
I have changed to Sequence, but I am wondering about the discussion about the `Sequence` type not including `np.ndarray` from here: https://github.com/Qiskit/qiskit-terra/pull/8812#discussion_r985427636

Reply:
Even though it says `Sequence` the primitives will keep supporting `np.ndarray`. Some people suggest moving back to `Iterable`, however these can actually be infinite (e.g. `itertools.count`), which raises concerns of secutiry, memory usage, and potential DOS attacks among others.

An alternative that I have proposed in the past is to move to `Collection` which does include `np.ndarray` but without being infinite. I personally like this approach, even though `Mapping`s are included so a dictionary could be passed.

For now, we are sticking to `Sequence` in the primitives since it is the least general. That way, generalizing it later on in the future will not be a breaking a change.